File: src/app/test/gmtime.py
import time
import random
import calendar

# Assume that the gmtime in libc supportes the most of time stamps in this range.
# The expected proportion of invalid cases is approximately 11% to 13%.
OFFSET = 2**53
N = 5000

precases = [
    time.strptime(
        "Jan 1 12:00:00 AM UTC 1967",
        "%b %d %I:%M:%S %p %Z %Y"),
    time.strptime(
        "Jan 9 11:45:14 AM UTC 0001",
        "%b %d %I:%M:%S %p %Z %Y"),
]

INT_MIN = -1 - 0x7fffffff
INT_MAX = 0x7fffffff
TS_LEAP_Y = 31622400
minv = INT_MIN * TS_LEAP_Y - OFFSET
maxv = INT_MAX * TS_LEAP_Y + OFFSET

def fmt(stamp, valid, tm: time.struct_time):
    # leap second is not allowed
    assert tm is None or tm.tm_sec < 60
    if not valid:
        return f"{{{stamp}, 0, {{ 0 }} }},"
    wday = (tm.tm_wday + 1) % 7
    return (
        f"{{{stamp}, 1, "
        f"{{ {tm.tm_sec}, {tm.tm_min}, {tm.tm_hour}, "
        f"{tm.tm_mday}, {tm.tm_mon - 1}, {tm.tm_year - 1900}, "
        f"{wday}, {tm.tm_yday - 1} }} }},"
    )

def gen_cases():
    for tm in precases:
        yield calendar.timegm(tm), True, tm

    for _ in range(N):
        stamp = random.randrange(minv, maxv)
        try:
            tm = time.gmtime(stamp)
            yield stamp, True, tm
        except (ValueError, OSError):
            yield stamp, False, None

def main():
    cases = list(gen_cases())
    inv = sum(not v for _, v, _ in cases)

    for c in cases:
        print(fmt(*c))

    print(f"// random [{minv:e}, {maxv:e})")
    print(f"// number of cases: {len(cases)}")
    print(f"// P(invalid cases) = {inv/len(cases):.8f}")


if __name__ == "__main__":
    main()

File: src/app/test/gmtime.c
#include <time.h>
#include <stdio.h>
#include <stdlib.h>

struct test
{
    time_t stamp;
    int valid;
    struct tm tm;
} tests[] = {
    #include "auto/gmtime.h"
};

#define len(x) (sizeof(x) / sizeof(x[0]))

static inline int eq(const struct tm *t1, const struct tm *t2)
{
    return t1->tm_year == t2->tm_year && t1->tm_mon == t2->tm_mon &&
           t1->tm_mday == t2->tm_mday && t1->tm_hour == t2->tm_hour &&
           t1->tm_min == t2->tm_min && t1->tm_sec == t2->tm_sec;
}

int main(int argc, char *argv[])
{
    putenv("TZ=UTC");
    for (int i = 0; i < len(tests); i++) {
        struct tm *r = gmtime(&tests[i].stamp);
        struct tm *e = &tests[i].tm;
        if (r == NULL && !tests[i].valid)
            continue;
        if (!eq(r, e)) {
            printf("failed %d: \n", i);
            printf("  result %s", asctime(r));
            printf("  expect %s", asctime(e));
            return 1;
        }
    }
    printf("all test cases (%d) passed!\n", (int)len(tests));
    return 0;
}
